Accelerate diskless master connections, and general re-connections

oranagra opened the pull request:
Diskless master has some inherent latencies.
1) fork starts with delay from cron rather than immediately
2) replica is put online only after an ACK. but the ACK
   was sent only once a second.
3) but even if it would arrive immediately, it will not
   register in case cron didn't yet detect that the fork is done.

Besides that, when a replica disconnects, it doesn't immediately
attempts to re-connect, it waits for replication cron (one per second).
in case it was already online, it may be important to try to re-connect
as soon as possible, so that the backlog at the master doesn't vanish.

In case it disconnected during rdb transfer, one can argue that it's
not very important to re-connect immediately, but this is needed for the
"diskless loading short read" test to be able to run 100 iterations in 5
seconds, rather than 3 (waiting for replication cron re-connection)

changes in this commit:
1) sync command starts a fork immediately if no sync_delay is configured
2) replica sends REPLCONF ACK when done reading the rdb (rather than on 1s cron)
3) when replconf ACK arrives, check check if the fork child terminated already
4) when a replica unexpectedly disconnets, it immediately tries to
   re-connect rather than waiting 1s
5) when when a child exits, if there is another replica waiting, we spawn a new
   one right away, instead of waiting for 1s replicationCron.

oranagra commented:
@antirez this is for you to decide if you want it (or parts of it) or not.
i realize that fast re-connection attempts can mean redis may be used for DDoS attacks. maybe we want parts of this (`cancelReplicationHandshake` specifically) configurable (so that they're only used by the test suite).

oranagra commented:
p.s. although this fix is enough for the short read test, there are still cases where the first ACK arrives before the fork finished.
Specifically this happens sometimes when the database is empty, i.e. you set up an empty database with master and replica, and the replica will become online with a 1 second delay since the rdb file reading is quick, and the first ACK is received by the master before the fork child terminated.

the only way i can see to solve this (if we want) is to use a pipe between the parent and child, so that when the child is done, it writes to the pipe, the parent wakes up, and writes the final EOF marker to the sockets, so when the ACK is received, it can immediately mark the replica as online, and not wait 1 second for the next ACK.

oranagra commented:
@antirez i rebased this PR, please have a look

antirez commented:
Thanks @oranagra, I need to account for some dedicated time in order to merge this, now I've a list of priorities, however I'm putting this into the Redis 6 milestone so that it can be merged, but possibly after RC1 because of time concerns.

antirez commented:
Rescheduled for 6.0.1.

madolson commented on src/replication.c:
```diff
@@ -2359,6 +2381,14 @@ void replicationHandleMasterDisconnection(void) {
+     * waiting 1 second may risk backlog being recycled. */
+    serverLog(LL_NOTICE,"Reconnecting to MASTER %s:%d",
+        server.masterhost, server.masterport);
+    if (connectWithMaster() == C_OK) {
+        serverLog(LL_NOTICE,"MASTER <-> REPLICA sync started");
+    }
```
The anti-code duplication in me thinks we should just move the log into connectWithMaster(), and have it return nothing. Although this part of the code is kind of all over the place.

oranagra replied:
thanks. i'll promote this to a function, and also consider calling it from `replicationSetMaster` so that when the user changed the configured master we don't wait 1 second to connect to it.

oranagra commented on src/replication.c:
```diff
@@ -825,7 +827,12 @@ void replconfCommand(client *c) {
+             * bgsave is done (since that depends on cron ticks), so run a
+             * quick check first (instead of waiting for the next ACK. */
+            if (server.rdb_child_pid != -1 && c->replstate == SLAVE_STATE_WAIT_BGSAVE_END)
+                checkChildrenDone();
```
This PR is from before the TLS changes, so this specific change might not be needed now (we reap the child as soon as the pipe close event arrives).

we should either also let it have a chance to check if the child pipe was closed (if the ack arrived before detecting the pipe close event) and let it clear server.rdb_pipe_conns, or drop this change if no longer needed.

if we see that there is a chance the replica sent an ACK (after seeing the EOF marker) before we detected the pipe was closed, we must handle it, otherwise we'll wait 1 second for the next ack.

oranagra commented:
@madolson @yossigo please have a quick look at my second commit (and the commit comment).
the first one was rebased.

specifically the concern about any undesired side effect for the change in `replicationSetMaster`

oranagra commented:
TODO:
- updateSlavesWaitingBgsave no longer needs to call startBgsaveForReplication

oranagra commented:
@yossigo i squashed my previous commits, rebased, and added another commit, please review it.
turns out that after your fix to the cluster tests the change i made to `replicationSetMaster` to immediately initiate the connection seems good, so i kept it.

oranagra commented:
@yossigo please review last two commits.
tests fail because of https://github.com/redis/redis/pull/7584

oranagra commented:
@madolson what do you mean by move to the other PR?
This is the only PR on this subject (acceleration).
the other linked PRs are ones that fix other bugs that this acceleration exposed.

p.s. since your previous review, the main change which i'm concerned about is that `replicationSetMaster` immediately initiate the connection (may have side effects). i.e. similar to the ones fixed 421def40c39cdb0bda20996ffbfc6cbbd9ec43bd

oranagra commented:
pushed a rebase from unstable to resolve failing test.

@madolson now i understand what you mean by the other PR. by mistake it included the contents of this PR (meant to include just the test fix). sorry for the mess.

yossigo commented on src/replication.c:
```diff
@@ -2476,17 +2485,22 @@ void replicationSetMaster(char *ip, int port) {
     }
     disconnectAllBlockedClients(); /* Clients blocked in master, now slave. */
 
+    /* Setting masterhost only after the call to freeClient since it calls
```
@oranagra So maybe it's better to set it to `NULL`, call `freeClient()` to clean up, then set the new masterhost and reconnect?

oranagra replied:
@yossigo that's what i did.
or am i missing something?

guybe7 commented:
:tada:
